### PR TITLE
Hanfei/apply

### PR DESCRIPTION
**Pull request by hanfei19910905**
apply represents correlated execution.

### CODE REVIEW

**Comment by hanfei19910905** [comment]
@coocood @zimulala @shenli PTAL

File: executor/builder.go [hunk 1]
```diff
@@ -119,6 +119,12 @@ func (b *executorBuilder) build(p plan.Plan) Executor {
 		return b.buildProjection(v)
 	case *plan.NewTableScan:
 		return b.buildNewTableScan(v)
+	case *plan.Apply:
+		return b.buildApply(v)
+	case *plan.Exists:
+		return b.buildExists(v)
+	case *plan.Max1Row:
```
Comment (shenli) [inline executor/builder.go]: Max1Row -> MaxOneRow

File: executor/executor.go [hunk 2]
```diff
@@ -97,6 +98,12 @@ type Executor interface {
 	Schema() expression.Schema
 }
 
+// NewExecutor executes a query.
+type NewExecutor interface {
+	Executor
+	Init()
```
Comment (shenli) [inline executor/executor.go]: Add comment for Init()

File: executor/new_executor.go [hunk 3]
```diff
@@ -676,3 +718,145 @@ func (e *NewSortExec) Next() (*Row, error) {
...
+	e.outerExec.Init()
+	for _, col := range e.outerSchema {
+		idx := col.Index
+		col.SetValue(&srcRow.Data[idx])
+	}
+	outerRow, err := e.outerExec.Next()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	srcRow.Data = append(srcRow.Data, outerRow.Data...)
```
Comment (shenli) [inline executor/new_executor.go]: Should you consider the handle in outerRow?

File: executor/new_executor.go [hunk 4]
```diff
@@ -676,3 +718,145 @@ func (e *NewSortExec) Next() (*Row, error) {
...
+func (e *ExistsExec) Next() (*Row, error) {
+	if !e.evaluated {
+		e.evaluated = true
+		srcRow, err := e.Src.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return &Row{Data: []types.Datum{types.NewDatum(srcRow != nil)}}, nil
+	}
+	return &Row{Data: []types.Datum{types.NewDatum(nil)}}, nil
```
Comment (shenli) [inline executor/new_executor.go]: You can return &Row{}, nil

File: executor/new_executor.go [hunk 5]
```diff
@@ -676,3 +718,145 @@ func (e *NewSortExec) Next() (*Row, error) {
...
+}
+
+// Max1RowExec represents max1row executor.
+type Max1RowExec struct {
```
Comment (shenli) [inline executor/new_executor.go]: It seems that this is similar with ExistsExec?

File: executor/new_executor.go (hunk 3 above)
Comment (hanfei19910905) [inline executor/new_executor.go]: no, I needn't

File: executor/new_executor.go (hunk 5 above)
Comment (hanfei19910905) [inline executor/new_executor.go]: max1row means when it returns more than one rows, throw exception.

Comment (coocood) [inline executor/new_executor.go]: Please add this in comment.

File: plan/plan.go [hunk 6]
```diff
@@ -58,6 +58,8 @@ type Plan interface {
 	GetSchema() expression.Schema
 	// Get ID.
 	GetID() string
+	// Judge weather this plan is correlated or not.
```
Comment (coocood) [inline plan/plan.go]: Check if this plan is correlated.

File: plan/expression_rewriter.go [hunk 7]
```diff
@@ -3,32 +3,58 @@ package plan
...
+	correlated bool
+}
+
+func (er *expressionRewriter) buildSubqueries(subq *ast.SubqueryExpr) (Plan, expression.Schema) {
+	if len(er.b.outerSchema) > 0 {
+		er.err = errors.New("Nested subqueries is not supported.")
+		return nil, nil
+	}
+	outerSchema := er.schema.DeepCopy()
+	for _, s := range outerSchema {
```
Comment (coocood) [inline plan/expression_rewriter.go]: element in a schema is column, should name it 'column'

File: plan/expression_rewriter.go [hunk 8]
```diff
@@ -3,32 +3,58 @@ package plan
...
+		s.Correlated = true
+	}
+	nb := &planBuilder{id: er.b.id, outerSchema: outerSchema}
```
Comment (coocood) [inline plan/expression_rewriter.go]: put outerSchemas in a stack in planBuilder, so we don't need to create a new planbuilder to build subquery.

File: plan/column_pruning.go [hunk 9]
```diff
@@ -49,9 +51,9 @@ func makeUsedList(usedCols []*expression.Column, schema expression.Schema) []boo
 }
 
 // PruneColumnsAndResolveIndices prunes unused columns and resolves index for columns.
```
Comment (coocood) [inline plan/column_pruning.go]: Add comments about the returned value.

File: executor/new_executor.go [hunk 10]
```diff
@@ -676,3 +718,145 @@ func (e *NewSortExec) Next() (*Row, error) {
 func (e *NewSortExec) Close() error {
 	return e.Src.Close()
 }
+
+// ApplyExec represents apply executor.
+type ApplyExec struct {
+	schema      expression.Schema
+	Src         NewExecutor
+	outerSchema expression.Schema
+	outerExec   NewExecutor
```
Comment (coocood) [inline executor/new_executor.go]: rename outer to inner.

File: plan/new_plans.go [hunk 11]
```diff
@@ -69,6 +69,24 @@ type Selection struct {
 	Conditions []expression.Expression
 }
 
+// Apply means correlated execution.
```
Comment (coocood) [inline plan/new_plans.go]: Please explain Apply in detail.

File: plan/expression_rewriter.go [hunk 12]
```diff
@@ -146,12 +146,17 @@ func (er *expressionRewriter) Leave(inNode ast.Node) (retNode ast.Node, ok bool)
 			return retNode, false
 		}
 		if column == nil {
-			column, err = er.b.outerSchema.FindColumn(v)
-			if err != nil {
-				er.err = errors.Trace(err)
-				return retNode, false
+			for _, outer := range er.b.outerSchemas {
```
Comment (coocood) [inline plan/expression_rewriter.go]: should iterate in reverse order.

File: plan/column_pruning.go [hunk 13]
```diff
@@ -49,9 +51,10 @@ func makeUsedList(usedCols []*expression.Column, schema expression.Schema) []boo
...
-func PruneColumnsAndResolveIndices(p Plan, usedCols []*expression.Column) ([]bool, error) {
+// This function returns a bool slice representing used columns, a column slice representing outer columns and an error.
+func PruneColumnsAndResolveIndices(p Plan, usedCols []*expression.Column) ([]bool, []*expression.Column, error) {
```
Comment (coocood) [inline plan/column_pruning.go]: rename usedCols to parentUsedCols would be better.

File: plan/column_pruning.go [hunk 14]
```diff
@@ -95,12 +98,43 @@ func PruneColumnsAndResolveIndices(p Plan, usedCols []*expression.Column) ([]boo
...
-				return nil, errors.Trace(err)
+				return nil, nil, errors.Trace(err)
 			}
 		}
 
 		v.schema.InitIndices()
-		return used, nil
+		return used, append(outer, outerCols...), nil
+	case *Apply:
+		_, outer, err := PruneColumnsAndResolveIndices(v.InnerPlan, v.InnerPlan.GetSchema())
```
Comment (coocood) [inline plan/column_pruning.go]: extract this block into a function and add more comment.

File: plan/new_plans.go [hunk 15]
```diff
@@ -69,6 +69,24 @@ type Selection struct {
...
+type Apply struct {
+	basePlan
+
+	InnerPlan   Plan
+	OuterSchema expression.Schema
+}
+
+// Exists check if a query returns result.
```
Comment (shenli) [inline plan/new_plans.go]: checks

File: executor/executor.go [hunk 16]
```diff
@@ -1363,3 +1372,18 @@ func (e *ReverseExec) Next() (*Row, error) {
 func (e *ReverseExec) Close() error {
 	return e.Src.Close()
 }
+
+func init() {
+	plan.EvalSubquery = func(p plan.Plan, is infoschema.InfoSchema, ctx context.Context) (d types.Datum, err error) {
```
Comment (shenli) [inline executor/executor.go]: There may be multiple columns in the result.

File: executor/new_executor.go [hunk 17]
```diff
@@ -248,19 +257,29 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
 	GroupByItems      []expression.Expression
 }
 
+// Init implements NewExecutor Init interface.
+func (e *AggregationExec) Init() {
+	e.Src.Init()
+	e.executed = false
+	e.groups = nil
+	e.currentGroupIndex = 0
+	for _, agg := range e.AggFuncs {
```
Comment (shenli) [inline executor/new_executor.go]: This is called in Executor.Close().

File: plan/column_pruning.go (hunk 13 above)
Comment (shenli) [inline plan/column_pruning.go]: Why make this public?

Comment (hanfei19910905) [inline plan/column_pruning.go]: plan package belonged to optimizer package long long ago.

File: executor/new_executor.go [hunk 18]
```diff
@@ -248,19 +257,29 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
 	groupMap          map[string]bool
 	groups            [][]byte
 	currentGroupIndex int
 	GroupByItems      []expression.Expression
 }
 
+// Init implements NewExecutor Init interface.
+func (e *AggregationExec) Init() {
+	e.Src.Init()
+	e.executed = false
```
Comment (shenli) [inline executor/new_executor.go]: The default value of bool variable is false.

Comment (hanfei19910905) [inline executor/new_executor.go]: We will init it multiple times.

File: plan/column_pruning.go [hunk 19]
```diff
@@ -165,39 +170,43 @@ func PruneColumnsAndResolveIndices(p Plan, usedCols []*expression.Column) ([]boo
...
 		v.schema.InitIndices()
-		return used, nil
-	case *Limit:
-		used, err := PruneColumnsAndResolveIndices(p.GetChildByIndex(0), usedCols)
-		return used, errors.Trace(err)
+		return used, nil, nil
+	case *Limit, *MaxOneRow:
+		used, outer, err := pruneColumnsAndResolveIndices(p.GetChildByIndex(0), parentUsedCols)
+		return used, outer, errors.Trace(err)
+	case *Exists:
```
Comment (shenli) [inline plan/column_pruning.go]: Why this is different from MaxOneRow?

**Comment by coocood** [comment]
LGTM

File: plan/column_pruning.go [hunk 20]
```diff
@@ -49,14 +51,15 @@ func makeUsedList(usedCols []*expression.Column, schema expression.Schema) []boo
 }
 
 // PruneColumnsAndResolveIndices prunes unused columns and resolves index for columns.
-func PruneColumnsAndResolveIndices(p Plan, usedCols []*expression.Column) ([]bool, error) {
+// This function returns a bool slice representing used columns, a column slice representing outer columns and an error.
+func pruneColumnsAndResolveIndices(p Plan, parentUsedCols []*expression.Column) ([]bool, []*expression.Column, error) {
 	//TODO: Currently we only implement index resolving, column pruning will be implemented later.
```
Comment (shenli) [inline plan/column_pruning.go]: Update this comment? I see you prune column at line 65.

Comment (hanfei19910905) [inline plan/column_pruning.go]: yeah, i will remove it

File: plan/column_pruning.go [hunk 21]
```diff
@@ -165,39 +170,43 @@ func PruneColumnsAndResolveIndices(p Plan, usedCols []*expression.Column) ([]boo
 					newSchema = append(newSchema, schema[i])
 				}
 			}
-			_, err := PruneColumnsAndResolveIndices(child, newSchema)
+			_, outer, err := pruneColumnsAndResolveIndices(child, newSchema)
```
Comment (shenli) [inline plan/column_pruning.go]: Why ignore the first return value?

File: plan/expression_rewriter.go [hunk 22]
```diff
@@ -3,32 +3,58 @@ package plan
...
+	ctxStack   []expression.Expression
+	p          Plan
+	schema     expression.Schema
+	err        error
+	aggrMap    map[*ast.AggregateFuncExpr]int
+	b          *planBuilder
+	correlated bool
+}
+
+func (er *expressionRewriter) buildSubqueries(subq *ast.SubqueryExpr) (Plan, expression.Schema) {
```
Comment (shenli) [inline plan/expression_rewriter.go]: buildSubqueries -> buildSubquery

**Comment by shenli** [comment]
LGTM

**Pull request merged**